### PR TITLE
Fix cuda cache config not being set correct

### PR DESCRIPTION
**Pull request by crtrott**
A bug introduced in 3.3.0
Thanks to Stan Moore for identifying it!

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp [hunk 1]
```diff
@@ -158,7 +158,7 @@ inline void check_shmem_request(CudaInternal const* cuda_instance, int shmem) {
   }
 }
 
-template <class KernelFuncPtr>
+template <class DriverType, class LaunchBounds, class KernelFuncPtr>
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:161]: I don't understand what the two template parameters you added are here for.  They are not being used.

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:161]: they are there so the static variable is unique for every combo of DriverType and LaunchBounds. KernelFuncPtr is often just void (*)()

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:161]: I see.  Please comment

File: core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp [hunk 2]
```diff
@@ -355,7 +355,9 @@ struct CudaParallelLaunchKernelInvoker<
 
     if (!Impl::is_empty_launch(grid, block)) {
       Impl::check_shmem_request(cuda_instance, shmem);
-      Impl::configure_shmem_preference(base_t::get_kernel_func(), prefer_shmem);
+      Impl::configure_shmem_preference<DriverType, LaunchBounds,
+                                       decltype(base_t::get_kernel_func())>(
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:359]: I don't think you need to specify the last template argument.  I expect FTAD does the right thing.

**Review by dalg24 — APPROVED** [review]

**Review by DavidPoliakoff — APPROVED** [review]

**Pull request merged**